File: src/main/java/ru/isb/bot/Bot.java
package ru.isb.bot;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.log4j.Log4j2;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.GetFile;
import org.telegram.telegrambots.meta.api.methods.ParseMode;
import org.telegram.telegrambots.meta.api.methods.send.SendChatAction;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.objects.Document;
import org.telegram.telegrambots.meta.api.objects.Update;
import ru.isb.bot.clients.NextcloudClient;
import ru.isb.bot.enums.Commands;
import ru.isb.bot.services.MessageServiceImpl;
import ru.isb.bot.utils.MessageUtils;

import static ru.isb.bot.services.MessageServiceImpl.MESSAGE_GPT_SPLIT;

@Log4j2
@Component
@Getter
@RequiredArgsConstructor
public class Bot extends TelegramLongPollingBot {

    @Value("${bot.name}")
    private String botUsername;
    @Value("${bot.token}")
    private String botToken;
    @Value("${bot.isb.chat.id}")
    private String ISB_CHAT_ID;

    private final MessageServiceImpl messageService;
    private final NextcloudClient nextcloudClient;

    @SneakyThrows
    @Override
    public void onUpdateReceived(Update update) {
        try {
            if (update.hasMessage()) {
                switchMessage(update);
            }
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            sendMessageException(e, update.getMessage().getChatId());
        }
    }

    @SneakyThrows
    private void switchMessage(Update update) {
        if (update.getMessage().hasText()) {
            handlerMessageText(update);
        } else if (update.getMessage().hasDocument()) {
            handlerMessageDocument(update);
        }
    }

    @SneakyThrows
    private void handlerMessageText(Update update) {
        switch (Commands.fromString(update.getMessage().getText())) {
            case SCHEDULE_GROUP, SCHEDULE -> executeMessages(
                    messageService.getSchedulesWeek(),
                    update.getMessage().getChatId()
            );
            case LIST_GROUP, LIST -> executeMessages(
                    messageService.getListGroup(),
                    update.getMessage().getChatId()
            );
            case ALL, ALL_GROUP, ALL_LINK -> executeMessages(
                    "@elektrik_gut @markin_ka @RA_prof @Mr_Ket1997 @Yureskii @Va1er1ev1ch @vladka_teb @polibuu @Desert567",
                    update.getMessage().getChatId()
            );
        }
        if (update.getMessage().getText().contains(MESSAGE_GPT_SPLIT)) {
            executeMessages(
                    getAnswerGPTMessageAndTyping(update),
                    update.getMessage().getChatId()
            );
        }
    }

    private String getAnswerGPTMessageAndTyping(Update update) {
        Thread typingThread = getTypingThread(update.getMessage().getChatId());
        typingThread.start();

        String answer = messageService.getAnswerGPTMessage(update.getMessage().getText());

        typingThread.stop();

        return answer;
    }
    @SneakyThrows
    private void handlerMessageDocument(Update update) {
        Document document = update.getMessage().getDocument();
        messageService.sendFileToNextcloud(
                downloadFile(
                        execute(new GetFile(document.getFileId())).getFilePath()
                ),
                document.getFileName()
        );
    }

    private Thread getTypingThread(Long chatId) {
        return new Thread(() -> {
            try {
                while (true) {
                    execute(new SendChatAction(String.valueOf(chatId), "TYPING", (int) Thread.currentThread().getId()));
                    Thread.sleep(2800);
                }
            } catch (Exception e) {
                log.error("Typing message error: {}", e.getMessage(), e);
            }
        });
    }

    private void executeMessages(String message, Long chatId) {
        MessageUtils.textSplitter(message).forEach(message4096 -> {
            try {
                execute(
                        sendMessage(
                                message4096,
                                chatId
                        )
                );
            } catch (Exception e) {
                log.error(e.getMessage(), e);
                sendMessageException(e, chatId);
            }
        });
    }

    private SendMessage sendMessage(String text, Long chatId) {
        SendMessage sendMessage = new SendMessage();
        sendMessage.setParseMode(ParseMode.HTML);
        sendMessage.setText(text);
        sendMessage.setChatId(chatId.toString());
        return sendMessage;
    }

    @SneakyThrows
    private void sendMessageException(Exception e, Long chatId) {
        execute(
                sendMessage(e.toString(), chatId)
        );
    }

}

File: src/main/java/ru/isb/bot/Bot.kt
package ru.isb.bot

import kotlinx.coroutines.*
import lombok.SneakyThrows
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.telegram.telegrambots.bots.TelegramLongPollingBot
import org.telegram.telegrambots.meta.api.methods.AnswerCallbackQuery
import org.telegram.telegrambots.meta.api.methods.GetFile
import org.telegram.telegrambots.meta.api.methods.ParseMode
import org.telegram.telegrambots.meta.api.methods.send.SendChatAction
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.objects.Update
import ru.isb.bot.enums.Commands
import ru.isb.bot.enums.Commands.Companion.fromString
import ru.isb.bot.services.MessageService
import ru.isb.bot.utils.MessageUtils
import ru.isb.bot.utils.MessageUtils.Companion.textSplitter
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.function.Consumer


@OptIn(DelicateCoroutinesApi::class)
@Component
class Bot(
    private val messageService: MessageService
) : TelegramLongPollingBot(), Logging {

    @Value("\${bot.token}")
    private val botToken: String = ""

    @Value("\${bot.name}")
    private val botUsername: String = ""

    @Value("\${bot.isb.chat.id}")
    private val ISB_CHAT_ID: String = ""

    private var gptEnableMap = HashMap<Long, Boolean>();

    override fun getBotUsername(): String = botUsername
    override fun getBotToken(): String = botToken

    private var job: Job? = null;

    @SneakyThrows
    override fun onUpdateReceived(update: Update) {
        val executor = Executors.newCachedThreadPool()
        val future = CompletableFuture<Void>();
        executor.execute() {
            future.thenRunAsync {
                asyncOnUpdateReceived(update)
            }
        }

        future.complete(null)
        executor.shutdown()
    }


    private fun asyncOnUpdateReceived(update: Update) {
        try {
            if (update.hasMessage()) {
                switchMessage(update)
            }
        } catch (e: Exception) {
            logger.error("Error onUpdateReceived ${e.message}", e)
            sendMessageException(e, update.message.chatId)
        } finally {
            job?.cancel()
        }
    }

    @SneakyThrows
    private fun switchMessage(update: Update) {
        if (update.message.hasText()) {
            handlerMessageText(update)
        } else if (update.message.hasDocument()) {
            handlerMessageDocument(update)
        }
    }

    @SneakyThrows
    private fun handlerMessageText(update: Update) {
        when (fromString(update.message.text)) {

            Commands.SCHEDULE_GROUP, Commands.SCHEDULE -> {
                job = GlobalScope.launch { typing(update.message.chatId) }
                executeMessages(
                    messageService.getSchedulesWeek(),
                    update
                )
            }

            Commands.LIST_GROUP, Commands.LIST -> {
                job = GlobalScope.launch { typing(update.message.chatId) }
                executeMessages(
                    messageService.getListGroup(),
                    update
                )
            }

            Commands.ALL, Commands.ALL_GROUP, Commands.ALL_LINK -> {
                job = GlobalScope.launch { typing(update.message.chatId) }
                executeMessages(
                "@kuzya\\_ram @markin\\_ka @RA\\_prof @Mr\\_Ket1997 @Yureskii @V0xP0puli @vladka\\_teb @polibuu @Desert567",
                    update
                )
            }

            Commands.HELP -> {
                job = GlobalScope.launch { typing(update.message.chatId) }
                executeMessages(
                    """
                        /schedule - Расписание занятий
                        /list - Список группы
                        /all - Вызвать всех участников группы
                        /help - Пояснение команд
                        /gpt - Команда включения режима GPT
                    """.trimIndent(),
                    update
                )
            }

            Commands.GPT -> {
                executeMessages(
                    """
                        ${if (switchingGptMode(update)) "Включен GPT режим. Бот будет" else "Отключен GPT режим. Бот не будет"} отвечать на все сообщения.
                    """.trimIndent(),
                    update
                )
            }

            else -> {
              if (gptEnableMap[update.message.chatId] == true) {
                    job = GlobalScope.launch { typing(update.message.chatId) }
                    executeMessages(
                        messageService.getAnswerGPTMessage(MessageUtils.concatInputMessage(update.message.text)),
                        update
                    )
              } else {}
            }
        }
    }

    @SneakyThrows
    private fun handlerMessageDocument(update: Update) {
        val document = update.message.document
        messageService.sendFileToNextcloud(
            downloadFile(
                execute(GetFile(document.fileId)).filePath
            ),
            document.fileName
        )
    }

    private suspend fun typing(chatId: Long) {
        try {
            while (job?.isActive == true) {
                execute(
                    SendChatAction(
                        chatId.toString(),
                        "TYPING",
                        Thread.currentThread().id.toInt()
                    )
                )
                delay(3000)
            }
        } catch (e: Exception) {
            logger.error("Typing message error: ${e.message}", e)
        }
    }

    private fun executeMessages(message: String, update: Update) {
        textSplitter(message).forEach(Consumer { message4096: String ->
            try {
                execute(
                    sendMessage(
                        message4096,
                        update.message.chatId
                    )
                )
            } catch (e: Exception) {
                logger.info("Info executeMessages: $message4096")
                logger.error(e.message ?: "Error executeMessages", e)
                sendMessageException(e, update.message.chatId)
            }
        })
    }

    private fun sendMessage(text: String, chatId: Long): SendMessage {
        val sendMessage = SendMessage()
        sendMessage.parseMode = ParseMode.MARKDOWN
        sendMessage.text = text
        sendMessage.chatId = chatId.toString()
        return sendMessage
    }

    @SneakyThrows
    private fun sendMessageException(e: Exception, chatId: Long) {
        execute(
            sendMessage(e.toString(), chatId)
        )
    }

    private fun switchingGptMode(update: Update) : Boolean {
        val chatId = update.message.chatId;
        val enableGpt = gptEnableMap[chatId]?:false;

        gptEnableMap.put(chatId, !enableGpt)
        return !enableGpt
    }

}